runtime/userspace/api/caliptra-api/src/image_loading/pldm_fdops.rs

```
@@ -0,0 +1,243 @@
+                )
+            };
+
+            ctx.last_requested_length = length;
+            (ctx.current_offset, length)
+        });
+
+        Ok((offset, request_length))
+    }
+
+    async fn download_fw_data(
+        &self,
+        offset: usize,
+        data: &[u8],
+        _component: &FirmwareComponent,
+    ) -> Result<TransferResult, FdOpsError> {
+        self.copy_data_to_buffer(offset, data).await?;
+        // update self.download_ctx
+        let should_yield = DOWNLOAD_CTX.lock(|ctx| {
+            let mut ctx = ctx.borrow_mut();
+            if ctx.total_downloaded >= ctx.total_length {
+                PLDM_STATE.lock(|state| {
+                    let mut state = state.borrow_mut();
+                    if *state == State::DownloadingHeader {
+                        *state = State::HeaderDownloadComplete;
+                        return false;
+                    } else if *state == State::DownloadingToc {
+                        *state = State::TocDownloadComplete;
+                        return true;
+                    } else if *state == State::DownloadingImage {
+                        *state = State::ImageDownloadComplete;
+                        return true;
+                    }
+                    false
+                })
+            } else {
+                ctx.current_offset += data.len();
+                false
+            }
+        });
+
+        if should_yield {
+            IMAGE_LOADING_TASK_YIELD.signal(());
+            PLDM_TASK_YIELD.wait().await;
+        }
+
+        Ok(TransferResult::TransferSuccess)
+    }
+
+    async fn is_download_complete(&self, _component: &FirmwareComponent) -> bool {
+        DOWNLOAD_CTX.lock(|ctx| ctx.borrow().download_complete)
+    }
+
+    async fn verify(
+        &self,
+        _component: &FirmwareComponent,
+        progress_percent: &mut ProgressPercent,
+    ) -> Result<VerifyResult, FdOpsError> {
+        *progress_percent = ProgressPercent::new(100).unwrap();
```

Should we set the verify progress (`Done`) based on the completion of authorization?

---

We also need to report the failure of verification. We may include download, verify context in `struct StreamingFdOps` .

---

For streaming boot, the images are verified during DOWNLOAD state. And since this verify function is called when the device is in the VERIFY PLDM state (after DOWNLOAD state), at this stage, all images have already been verified, so the progress is always 100%.

---

Yes we can report here the verify result. i.e. if any of the images have failed verification, then we can report it here as failure. I can add the verification result in the download ctx.